components/feature/media/src/main/java/mozilla/components/feature/media/middleware/LastMediaAccessMiddleware.kt

```diff
@@ -4,34 +4,196 @@
+            }
+            is TabListAction.RemoveAllPrivateTabsAction -> {
+                context.state.privateTabs.forEach { mediaTabsStatus.remove(it.id) }
+            }
+        }
+
         next(action)
 
-        if (action is MediaSessionAction.UpdateMediaPlaybackStateAction &&
-            action.playbackState == MediaSession.PlaybackState.PLAYING
-        ) {
-            context.dispatch(LastAccessAction.UpdateLastMediaAccessAction(action.tabId))
-        } else if (action is MediaSessionAction.DeactivatedMediaSessionAction) {
-            context.dispatch(LastAccessAction.UpdateLastMediaAccessAction(action.tabId, 0))
+        when (action) {
+            is MediaSessionAction.UpdateMediaPlaybackStateAction -> {
+                if (action.playbackState == MediaSession.PlaybackState.PLAYING) {
+                    context.ifNormalTab(action.tabId) {
+                        val mediaStartedPlayingTimestamp = System.currentTimeMillis()
+
+                        context.dispatch(
+                            LastAccessAction.UpdateLastMediaAccessAction(
+                                it.id, mediaStartedPlayingTimestamp
+                            )
+                        )
+                        rememberTabStatusForMediaPlaying(it, mediaStartedPlayingTimestamp)
+                    }
+                }
+            }
+            // GeckoView could deactivate the media session in various circumstances.
+            // (eg: media started playing in other tabs).
+            // We know media is not playing anymore. But the user might still be on that page.
+            is MediaSessionAction.DeactivatedMediaSessionAction -> {
+                context.ifNormalTab(action.tabId) {
+                    maybeUpdateTabStatusForMediaEnded(it)
+                    maybeDispatchPlaybackEnded(it, context)
+                }
+            }
+            // User navigated to another URL.
+            // We know this is different than that of the page on which media started playing
+            // but might still be the same HTML document and media continues to play.
+            is ContentAction.UpdateUrlAction -> {
+                context.ifNormalTab(action.sessionId) {
+                    maybeUpdateTabStatusForUrlUpdated(it, action.url)
+                    maybeDispatchPlaybackEnded(it, context)
+                }
+            }
+            // Cleanup mediaTabsStatus if media tabs were closed. We need to be careful about
+            // not leaking state but more so about preventing it get uncontrollably big.
+            is TabListAction.RemoveTabAction -> {
+                mediaTabsStatus.remove(action.tabId)
+            }
+            is TabListAction.RemoveTabsAction -> {
+                action.tabIds.forEach { mediaTabsStatus.remove(it) }
+            }
+            is TabListAction.RemoveAllTabsAction -> {
+                mediaTabsStatus.clear()
+            }
```

Comment:
See above re: managing this new state. We wouldn't have to delete it here if we stored it in the tab.